statscache/app.py

```diff
@@ -39,16 +38,22 @@ def paginate(queryset):
     # parse view arguments
     page_number = int(flask.request.args.get('page', default=1))
     page_length = min(
-        MAXIMUM_ROWS_PER_PAGE,
+        maximum_rows_per_page,
         int(flask.request.args.get('rows_per_page',
-                                   default=DEFAULT_ROWS_PER_PAGE))
+                                   default=default_rows_per_page))
     )
 
-    items_count = queryset.count()
+    items_count = int(limit or queryset.count())
     page_count = items_count / page_length + \
         (1 if items_count % page_length > 0 else 0)
-    queryset = \
-        queryset.offset((page_number - 1) * page_length).limit(page_length)
+    page_start = (page_number - 1) * page_length
+    page_stop = min(page_length, items_count - page_start)
+    queryset = queryset.slice(page_start, page_stop)
+
+    if page_start > items_count:
+        # In this case, an empty response is safely generated, but it would be
+        # bad practice to respond to invalid requests as if they were correct.
+        flask.abort(400)
```

Comment:
It's a shame it's reserved for WebDav, but http code 422 would be neat to use here.  *c'est la vie*.